### PR TITLE
feat(*): allow to select auth providers + add multiworkspace with subdomain management

### PR DESCRIPTION
**Pull request by AMoreaux**
## Summary
Add support for multi-workspace feature and adjust configurations and states accordingly.
- Introduced new state isMultiWorkspaceEnabledState.
- Updated ClientConfigProviderEffect component to handle multi-workspace.
- Modified GraphQL schema and queries to include multi-workspace related configurations.
- Adjusted server environment variables and their respective documentation to support multi-workspace toggle.
- Updated server-side logic to handle new multi-workspace configurations and conditions.

### CODE REVIEW

**Review by martmull — COMMENTED** [review]
LGTM, one comment about default value

File: packages/twenty-server/src/engine/core-modules/environment/environment-variables.ts
Comment (AMoreaux) [reply packages/twenty-server/src/engine/core-modules/environment/environment-variables.ts:238]: Done 👌

**Comment by AMoreaux** [comment]
⚠️ Before merging it wait for the PR that adds subdomains.

**Comment by AMoreaux** [comment]
@martmull I converted this PR to a draft to avoid the merge since we must wait for the review on child PR to have the full feature.

File: packages/twenty-server/src/engine/core-modules/auth/services/sign-in-up.service.ts
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/auth/services/sign-in-up.service.ts:301]: workspacesCount

**Review by charlesBochet — APPROVED** [review]

File: packages/twenty-front/src/modules/auth/sign-in-up/components/SignInUpWorkspaceScopeForm.tsx
Comment (charlesBochet) [inline packages/twenty-front/src/modules/auth/sign-in-up/components/SignInUpWorkspaceScopeForm.tsx:28]: move to SignInUpWorkspaceScopeFormEffect

File: packages/twenty-front/src/modules/workspace/components/WorkspaceProviderEffect.tsx
Comment (charlesBochet) [inline packages/twenty-front/src/modules/workspace/components/WorkspaceProviderEffect.tsx:31]: see if can leverage react-helmet

File: packages/twenty-front/src/modules/auth/types/signInUpMode.type.ts
Comment (charlesBochet) [inline packages/twenty-front/src/modules/auth/types/signInUpMode.type.ts:1]: filenaming

File: packages/twenty-front/src/modules/settings/security/components/SettingsSSOIdentitiesProvidersListCard.tsx
Comment (charlesBochet) [inline packages/twenty-front/src/modules/settings/security/components/SettingsSSOIdentitiesProvidersListCard.tsx:17]: no .state

File: packages/twenty-front/src/modules/settings/security/components/SettingsSecurityOptionsList.tsx
Comment (charlesBochet) [inline packages/twenty-front/src/modules/settings/security/components/SettingsSecurityOptionsList.tsx:35]: extract this to util so we can test it easily and we the typing complexity is "hidden"

File: packages/twenty-front/src/modules/settings/security/types/AuthProviders.type.ts
Comment (charlesBochet) [inline packages/twenty-front/src/modules/settings/security/types/AuthProviders.type.ts:1]: GOOGLE
MICROSOFT

isGoogleAuthEnabled
isMicrosoftAuthEnabled

map(authEnablekyey: AuthProvidersKeys): GOOGLE | MICROSOFT (

)

Comment (charlesBochet) [reply packages/twenty-front/src/modules/settings/security/types/AuthProviders.type.ts:1]: not mandatory but I would have "avoided the complexity like this"

File: packages/twenty-front/src/utils/workspace-url.helper.ts
Comment (charlesBochet) [inline packages/twenty-front/src/utils/workspace-url.helper.ts:51]: file naming + split into multi files

File: packages/twenty-server/src/engine/core-modules/auth/services/auth.service.ts
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/auth/services/auth.service.ts:122]: isAuthProviderEnabled

File: packages/twenty-server/src/engine/core-modules/auth/controllers/google-auth.controller.ts
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/auth/controllers/google-auth.controller.ts:61]: authProvider: 'google' | 'microsoft'

File: packages/twenty-server/src/engine/core-modules/auth/services/sign-in-up.service.ts
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/auth/services/sign-in-up.service.ts:200]: workspaceValidator.validateAuth(() => { isValid(authProvider), ${authProvider} not enabled)

File: packages/twenty-server/src/engine/core-modules/user/services/user.service.ts
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/user/services/user.service.ts:138]: give a more explicit name

Comment (charlesBochet) [reply packages/twenty-server/src/engine/core-modules/user/services/user.service.ts:138]: saveDefautWorkspaceIfUserHasAccessOrThrow

Comment (charlesBochet) [reply packages/twenty-server/src/engine/core-modules/user/services/user.service.ts:138]: or extract hasUserAccessToWorkspace

File: packages/twenty-server/src/engine/core-modules/user/user.validate.ts
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/user/user.validate.ts:4]: assertIsDefinedOrThrow

File: packages/twenty-server/src/engine/core-modules/user/user.validate.ts
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/user/user.validate.ts:27]: we would avoid typeof usage usually to lower complexity

File: packages/twenty-server/src/engine/core-modules/workspace-invitation/services/workspace-invitation.service.ts
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/workspace-invitation/services/workspace-invitation.service.ts:181]: move cast to an util

File: packages/twenty-server/src/engine/core-modules/workspace/workspace.resolver.ts
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/workspace/workspace.resolver.ts:156]: we could extract that to another resolver

File: packages/twenty-server/src/engine/core-modules/workspace/workspace.validate.ts
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/workspace/workspace.validate.ts:36]: see comments above

**Review by charlesBochet — APPROVED** [review]

File: packages/twenty-server/src/engine/core-modules/auth/services/sign-in-up.service.ts
Comment (zeckdude) [reply packages/twenty-server/src/engine/core-modules/auth/services/sign-in-up.service.ts]: What do you mean? Please expand.

Comment (zeckdude) [reply packages/twenty-server/src/engine/core-modules/auth/services/sign-in-up.service.ts]: @greptileai What do you mean? Please expand.

Comment (zeckdude) [reply packages/twenty-server/src/engine/core-modules/auth/services/sign-in-up.service.ts]: @greptile-apps What do you mean? Please expand.

**Pull request merged**
